### ISSUE TITLE
Using PBKDF2WithHmacSHA1 with 10k iterations and 48B output is too slow

### ISSUE THREAD

**Issue by cermakcz** [scottyab/secure-preferences#18]
I'd very much like to use this library, but the initialization of the key takes 3 seconds on my Nexus 5. Since this is done in Application's onCreate (because I need the settings there), it's not usable.

I understand that you're trying to make it as easy to use and as secure as possible, but would you consider allowing users of the library to provide their own key? Since the encryption is really more of an obfuscation (anyone can decompile the app and generate the same key), it wouldn't be a problem for me to have less security, but 100x faster startup time :)

Also here's some discussion about the slowness: http://stackoverflow.com/questions/24652602/pbkdf2withhmacsha1-key-generation-takes-too-long-on-android

**Comment by scottyab** [comment]
wow, 3 seconds, i didn't know it was that long. Maybe some kind of lite mode option would suitable or as you suggest providing own key could be good. I'll review that post.

**Comment by scottyab** [comment]
Hey @cermakcz Confirmed. I'm seeing init time of 4.4s on nexus 4. Let's make suppling your own key an enhancement. I'd welcome PR if you have time, otherwise I'd add to my todo.

**Comment by shaunidiot** [comment]
Can confirm. Lags on Asus Zenphone 2 and Samsung devices. Starting an activity and retrieving 2 strings from SecurePreferences takes about 7 seconds for the app to actually do something after that. Been using this library from v0.0.4 so I'm reverting back just for now (no hate for this version though!). Hopefully, something can be done about it! Good luck! And thanks! :+1:

**Comment by cermakcz** [comment]
Hey @scottyab, I started looking into it and so far it seems that the best way to do it would be creating another constructor which would take `AesCbcWithIntegrity.SecretKeys` instead of the password. Plus, it'd be nice to add another method to `AesCbcWithIntegrity`, namely `generateKeyFromPassword(String password, String salt, int iterationCount, int keyLength)`, which would allow easy creation of the `SecreyKeys` with whatever strenght the user wants. But I'm not sure if it wouldn't violate the purpose of the library - providing simple and secure encryption. What do you think?

**Comment by scottyab** [comment]
Great minds think alike :+1: , I've just added a new constructor to take `AesCbcWithIntegrity.SecretKeys` and I'm just writing a unit test for it. As for the changes to 'AesCbcWithIntegrity', personally it makes sense, it just gives more options the default are still stronger.

**Comment by cermakcz** [comment]
Thanks for the change, I just started working on it when I read that you've already done it :) I'll go and see if there's something I can do about the AesCbcWithIntegrity change.

**Comment by scottyab** [comment]
Thanks for that. Hoping @tozny is ok with and can merge the PR. I'll pull once accepted. In Secure Preferences I'm actually using my fork of [java-aes-crypto](https://github.com/scottyab/java-aes-crypto) as wanted to reference the dependancy from maven central.

**Comment by cermakcz** [comment]
Yeah I noticed. That's why I was hoping you'd pull it to your fork and make a new version, like you said.

**Comment by scottyab** [comment]
@cermakcz i've pulled your fork of java-aes-crypto into my fork added unit test and published as 0.0.3.  Just waiting on maven central to propagate, then I'll release a new version of secure prefs 0.1.3 dependent on `com.scottyab:aes-crypto:0.0.3`

**Comment by cermakcz** [comment]
@scottyab Cool, thanks.

**Comment by scottyab** [comment]
Fixed in 0.1.3 (once propagated to maven central) 

Example... 
```java
AesCbcWithIntegrity.SecretKeys myKey = AesCbcWithIntegrity.generateKeyFromPassword(Build.SERIAL,AesCbcWithIntegrity.generateSalt(),1000);
            SharedPreferences securePrefs1000 = new SecurePreferences(this, myKey, "my_prefs_1000.xml");
```

**Comment by mkaarthick** [comment]
Hi, i tried using password, but still app lags.

**Comment by scottyab** [comment]
@mkaarthick I'm not sure I understand your comment. You can manually pass in your own keys created with `generateKeyFromPassword` method that takes a int of the `iterationCount`.  To increase the speed (but reduce security) you can lower the iterationCount. 

```java
AesCbcWithIntegrity.SecretKeys mykeys = AesCbcWithIntegrity.generateKeyFromPassword(password, salt, iterationCount);
SecurePreferences securePrefs = new SecurePreferences(getContext(), mykeys, "pref-file");
```

**Comment by ShinJJang** [comment]
`handlePasswordChange()` also need iteration. All method is private static, so I can't override. If to find method not to build library, the only way is inheritance with `SecurePreference` and overwrite method `handlePasswordChange()`. But, this is not cool.

**Comment by grmaciel** [comment]
Unfortunately i released an app with that in my application onCreate and its a messy situation right now. I can't force the users to logout and login cause that's a bit expensive for us. Would be a way for me to use the same keu generated by the class and then storing to use to create my SecretKeys? Any idea on that?

Thank you,
